app/controllers/application_controller.rb

```diff
@@ -5,6 +5,129 @@ class ApplicationController < Sinatra::Base
   configure do
     set :public_folder, 'public'
     set :views, 'app/views'
+    enable :sessions
+    set :sessions_secret, "secret"
+  end
+  
+  get "/signup" do
+    if !logged_in?
+      erb :'users/signup'
+    else
+      redirect '/tweets'
+    end
+  end
+
+  post "/signup" do
+    if params[:username] == "" || params[:password] == "" || params[:email] == ""
+      redirect "/signup", :locals => {:message => "*** Please fill-in all fields. ***"}
+    elsif
+      @user = User.create(username: params[:username], password: params[:password], email: params[:email])
+      session[:id] = @user[:id]
+      redirect "/tweets"
+    else
+      erb :'users/signup', :locals => {:message => "*** Please try again. ***"}
+    end
+  end
+
+  get "/login" do
+    if !logged_in?
+      erb :'users/login'
+    else
+      redirect '/tweets'
+    end
+  end
+
+  post "/login" do
+    @user = User.find_by(username: params[:username])
+    if @user && @user.authenticate(params[:password])
+      session[:id] = @user[:id]
+      redirect "/tweets"
+    else
+      erb :'users/login', :locals => {:message => "*** The username and password provided do not match. Please try again. ***"} 
+    end
+  end
+
+  get "/logout" do
+    session.clear
+    redirect "/login"
+  end
+
+  get "/" do
+    redirect "/tweets"
+  end
+
+  get "/tweets" do
+    if logged_in?
+      @fweets = Tweet.all
+      @posts = []
+      @fweets.each do |f|
+        @posts << {User.find(f.user_id).username => f.content, user_id: f.user_id, id: f.id}
+      end
+      @user = User.find(session[:id])
+      erb :'tweets/index'
+    else redirect "/login"
+    end
+  end
+
+  get '/tweets/new' do
+    if logged_in?
+      erb :'/tweets/new'
+    else redirect "/login"
+    end
+  end
+
+  post '/new' do
+    if logged_in? && params[:content] != ""
+      @content = params[:content]
+      Tweet.create(content: @content, user_id: session[:id])
+    else redirect "/login"
+    end
+    redirect "/tweets"
+  end
+
+  get "/users/:slug" do
+      @user = User.find_by(username: params[:slug])
+      @fweets = Tweet.where(user_id: @user.id)
+      erb :'/users/slug'
+  end
+
+  post "/tweet/:id/delete" do
+    if logged_in?
+      if params[:delete]
+        Tweet.delete(params[:id])
+        redirect "/tweets"
+      end
+    else redirect "/login"
+    end
+  end
+
+  get "/tweets/:id/edit" do
+    @fweet = Tweet.find(params[:id])
+    if logged_in? && @fweet.user_id == session[:id]
```

Comment:
Prefer to find tweets that are already only associated to the `current_user` than to manually check the `user_id` foreign key of the tweet against the `current_user.id`

```ruby
tweet = current_user.tweets.find_by(:id => params[:id])
if tweet && tweet.destroy
  redirect "/tweets"
else
  redirect "/tweets/#{tweet.id}"
end
```

We're using the association collection method provided by [`has_many`](http://api.rubyonrails.org/classes/ActiveRecord/Associations/ClassMethods.html#method-i-has_many) to query the users tweets. 

Note the usage of [`find_by`](http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find-by) vs [`find`](http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find). Had we used `find` here, ActiveRecord would raise an error in the event that we could not find a tweet associated with this user. `find_by` will allow this action to fail gracefully simply returning nil if the user is trying to delete a tweet that they do not have access to.